### ISSUE TITLE
Write an xtrace/xtruss-like sample program

### ISSUE THREAD

**Issue by psychon** [psychon/x11rb#442]
If we add request parsing, we have all the necessary ingredients for doing something like `xtrace`.

To keep things simple
- one could at first limit to TCP connections (but I don't think that unix sockets would add much complications - I am just not sure how/if they are cleaned up)
- each connection could spawn two threads, one per "forwarding direction"
- auth info would not really be handled; instead, clients could connect unauthentication and "proper auth" would be added inline

**Comment by psychon** [comment]
A first version is quickly hacked together (this listens on `DISPLAY=:4` via tcp and forwards connections to `:0` via an unix socket; it works for me since my X server accepts all unix socket connections from processes running as my user):
```rust
use std::io::{Error as IOError, ErrorKind, Read, Write};
use std::net::{TcpListener, TcpStream};
use std::os::unix::net::UnixStream;
use std::thread;

fn forward_impl(mut read: impl Read, mut write: impl Write) -> Result<(), IOError> {
    let mut buffer = [0; 4096];
    loop {
        let amount = match read.read(&mut buffer) {
            Ok(amount) => amount,
            Err(ref e) if e.kind() == ErrorKind::WouldBlock => continue, // TODO: Use poll()
            Err(e) => return Err(e),
        };
        if amount == 0 {
            return Ok(());
        }
        match write.write_all(&buffer[..amount]) {
            Err(e) if e.kind() == ErrorKind::BrokenPipe => return Ok(()),
            Err(e) => return Err(e),
            Ok(()) => {}
        }
    }
}

fn forward(read: impl Read, write: impl Write) {
    forward_impl(read, write).unwrap();
}

fn handle_client(client: TcpStream) -> Result<(), IOError> {
    let client2 = client.try_clone()?;
    let server = UnixStream::connect("/tmp/.X11-unix/X0")?;
    let server2 = server.try_clone()?;

    client.set_nonblocking(true)?;
    server.set_nonblocking(true)?;

    thread::spawn(|| forward(client, server2));
    thread::spawn(|| forward(server, client2));
    Ok(())
}

fn main() -> Result<(), IOError> {
    let listener = TcpListener::bind("127.0.0.1:6004")?;
    loop {
        let (socket, _addr) = listener.accept()?;
        handle_client(socket)?;
    }
}
```
Future work includes fixing 100% CPU usage (busy loop trying to read on a nonblocking socket) and perhaps have better panic handling. But it is surprisingly simple to get this started.

(Also: I guess this should use `x11rb::rust_connection::Stream` so that FD-passing is supported and the `Poll` implementation from this crate can just be used.)

**Comment by psychon** [comment]
New version which can print setup and errors/events: https://gist.github.com/psychon/4f5916f91bde7b1a3a3a8fdc9e9ba5b9
(This has some bugs; i.e. it fails to parse "everything" if more than one X11 packet is read at once)

While writing this, it occurred to me that it would be nice if there was some generic way to parse a reply given the corresponding request...

**Comment by psychon** [comment]
I updated https://gist.github.com/psychon/4f5916f91bde7b1a3a3a8fdc9e9ba5b9. This now also parses requests. Plus, this keeps track of `QueryExtension` requests and replies so that it can properly decode requests.

Some observations:
- I want to change the API of `x11rb::protocol::Request::parse` to return `Result<(Self, &[u8]), _>` instead of just the `Self` in there.
- The X11 server sends "too long" `QueryExtensionReply`s:
```
Reply to something; parsing this requires more state: [1, 0, 5, 0, 0, 0, 0, 0, 1, 135, 85, 137, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
  Well, it's actually a QueryExtensionReply { response_type: 1, sequence: 5, length: 0, present: true, major_opcode: 135, first_event: 85, first_error: 137 } for ext XKEYBOARD with trailing [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
Reply to something; parsing this requires more state: [1, 1, 6, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
```
- Some generic API for parsing replies  is needed. Does it make sense to add this to the `Request` enum? Something like `fn reply_parser(&self) -> Option<???>`, where `None` is used for requests that do not have a reply.
  - The `???` must contain/remember the type of the reply to parse.
    - So, either we add lots of zero-sized types that implement something like `TryParse` for this, `???` would then by `dyn TryParse` or something like it (well, it needs to take a `&self` argument for parsing).
    - Or alternatively `???` becomes some kind of `Fn(&{u8]) -> Result<(???, &[u8]), ParseError>`.
  - The second "problem" is: what should be the result of parsing?
    - For `xtrace`, something like `dyn Debug` would be enough, but that's not really usable for much in general.
    - Should we add a new `Reply` enum similar to the existing `Request`, `Error`, and `Event` enums?
      - This could then have a method `fn try_parse(kind: Kind, reply: &[u8])` where `Kind` is a new enum for all the possible replies... hm, that's a bit ugly. How about `Kind = Request` and this returns `ParseError` (or perhaps a better, new error enum) for requests that do not have replies?

So much for a random dump of thoughts. Perhaps others have some suggestions? Is it even worse pursuing this direction further? Will anything besides the `xtrace` example need this?
CC @khuey (and I guess @eduardosm gets a notification for this anyway...)

**Comment by eduardosm** [comment]
> The X11 server sends "too long" `QueryExtensionReply`s:

It is just padding to 32 bytes, isn't it? (which does not appear in the XML definitions)

> Some generic API for parsing replies is needed. Does it make sense to add this to the `Request` enum? Something like `fn reply_parser(&self) -> Option<???>`, where `None` is used for requests that do not have a reply.

My idea would be to have a `Reply` enum whose parse function also takes the header of the request (`major_opcode`...) and a `ExtInfoProvider`.

**Comment by psychon** [comment]
> It is just padding to 32 bytes, isn't it? (which does not appear in the XML definitions)

Ah right. I was only thinking about "multiple of four" and forgot the 32 bytes "base size". Thanks.

**Comment by psychon** [comment]
I updated the gist. Problems include: errors when a client sends a request for an extension that was not enabled and thus cannot be parsed.

@khuey I had to write my own version of "parse a request's length field" again: https://gist.github.com/psychon/4f5916f91bde7b1a3a3a8fdc9e9ba5b9#file-xtrace-rs-L109-L115
Because I need to know how much bytes were parsed / where the next request begins. No idea how to get this info otherwise.
(I noticed this because a `PolyFillRectangle` request failed to parse. That request contains a list with an implicit length aka "everything until the end of the request". To parse this correctly, the slice that is given to the parsing function has to have the correct length.)

**Comment by khuey** [comment]
Why can't you do something like
```rust
let (header, remaining) = x11_utils::parse_request_header(&remaining[..], BigRequests::Enabled).unwrap();
let message_size = header.remaining_length as usize * 4;
if message_size > remaining.len() {
    return 0;
}
let (message_body, next_message) = remaining.split_at(message_size);
remaining = next_message;
let request = protocol::Request::parse(header, message_body, &mut vec![], &state.ext_info);
```

**Comment by psychon** [comment]
> let message_size = header.remaining_length as usize * 4;

That assumes that this is not a big requests. Big requests need a `+4` for the size, but the parser does not tell me if "this" was a big request or not. (In theory, one could even use big request encoding and then send a "small" request, i.e. having the high bytes of the length field be zero...)

The best idea I had so far is to get raw pointers and compute the difference between pointers, but... no, thanks.

**Comment by khuey** [comment]
The `remaining` buffer that comes out of `parse_request_header` has been adjusted for the header size (including whether or not it's a big request).

**Comment by psychon** [comment]
My problem is:
```rust
// This function should parse one request from the bytes and return the number of bytes that were consumed
fn parse_one_request(incoming: &[u8]) -> usize {
    let (header, remaining) = x11_utils::parse_request_header(&remaining[..], BigRequests::Enabled).unwrap();
    // Let's just consume the request without looking at its content.... so, what is its length?
    // Get a pointer to the end of the request:
    let end = remaining.as_ptr().add(remaining.len());
    // Compute the request length as "end - start". This does not actually compile and I have no idea how to do this nicely.
    end - incoming.as_ptr()
}
```

**Comment by khuey** [comment]
```rust
let message_length = header.remaining_length * 4;
if remaining.len() < message_length {
    return 0;
}
message_length + incoming.len() - remaining.len()
```
?

**Comment by psychon** [comment]
Ah, thanks. Yeah, that should work. `incoming.len() - remaining.len()` is the length of the header.

Now... meh, next problem. Partial reads. When I provide not enough bytes, I get a `ParseError`. Everywhere else I can work around this by knowing beforehand the minimum length. Here, the minimum length is four bytes (`NoOperation`), but with big requests I need at least eight bytes of data.

**Comment by khuey** [comment]
There are errors in `parse_request_header` that are not "I need more data". (Namely a request length that is too low for even the header.) Not sure what we should do about that.

**Comment by psychon** [comment]
Fixed by #509
